### PR TITLE
NeMo 2.0 documentation upgrade

### PR DESCRIPTION
**Pull request by dimapihtar**
# What does this PR do ?

Add a one line overview of what this PR aims to accomplish.

**Collection**: [Note which collection this PR will affect]

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: docs/source/features/mixed_precision.rst [hunk 1]
```diff
@@ -13,8 +13,49 @@ NeMo Framework supports half-precision FP16 and BF16 computation training via Me
 This training recipe uses half-precision in all layer computation keeping the model states (optimizer states and master parameters) in single-precision.
 To avoid repeated data type casting at each layer computation, Megatron Core keeps a separate copy of half-precision parameters that is updated after each optimizer step.
 
-Half-precision training is enabled when setting ``precision`` to either of ``fp16-mixed`` or ``bf16-mixed`` along with  ``megatron_amp_O2=true``.
-The parameter gradients are computed in the same half-precision, and the precision of gradient reduce-scatter across data-parallel GPUs can be set by ``optim.grad_sync_dtype``.
+Half-precision training is enabled when setting trainer's ``plugins`` to either of ``fp16-mixed`` or ``bf16-mixed``.
```
Comment (jgerh) [inline docs/source/features/mixed_precision.rst:16]: revise

Half-precision training is enabled when setting trainer's ``plugins`` to either ``fp16-mixed`` or ``bf16-mixed``.

File: docs/source/features/mixed_precision.rst [hunk 2]
```diff
@@ -13,8 +13,49 @@ NeMo Framework supports half-precision FP16 and BF16 computation training via Me
...
+The parameter gradients are computed in the same half-precision, and the precision of gradient reduce-scatter across data-parallel GPUs is set automatically according to the trainer's precision.
+
+Usage
```
Comment (jgerh) [inline docs/source/features/mixed_precision.rst:19]: suggested revision to an imperative verb

Utilize Half-Precision Training or Implement Half-Precision Training

File: docs/source/features/mixed_precision.rst [hunk 3]
```diff
@@ -13,8 +13,49 @@ NeMo Framework supports half-precision FP16 and BF16 computation training via Me
 This training recipe uses half-precision in all layer computation keeping the model states (optimizer states and master parameters) in single-precision.
 To avoid repeated data type casting at each layer computation, Megatron Core keeps a separate copy of half-precision parameters that is updated after each optimizer step.
 
-Half-precision training is enabled when setting ``precision`` to either of ``fp16-mixed`` or ``bf16-mixed`` along with  ``megatron_amp_O2=true``.
```
Comment (jgerh) [inline docs/source/features/mixed_precision.rst:16]: Line 9/9 fix capitalization in heading above

Half-Precision Training
===============

File: docs/source/features/mixed_precision.rst (hunk 2 above)
Comment (jgerh) [inline docs/source/features/mixed_precision.rst:19]: revise Lines 16-19  and delete Usage heading. Change this text like this:

Enable Half-Precision Training
^^^^^^^^^^^^^^^^^^^

You can enable half-precision training by setting the trainer's ``plugins`` to either of ``fp16-mixed`` or ``bf16-mixed``.
The parameter gradients are computed in the same half-precision, and the precision of gradient reduce-scatter across data-parallel GPUs is set automatically according to the trainer's precision.

File: docs/source/features/mixed_precision.rst [hunk 4]
```diff
@@ -13,8 +13,49 @@ NeMo Framework supports half-precision FP16 and BF16 computation training via Me
...
+  from nemo.collections.llm.recipes.precision.mixed_precision import bf16_mixed, fp16_mixed
+
+  # Load recipe
+  recipe = partial(llm.llama3_8b.pretrain_recipe)()
+
+  # Change precision
+  recipe.trainer.plugins = fp16_mixed()
 
 FP8 Training
 ============
```
Comment (jgerh) [inline docs/source/features/mixed_precision.rst:61]: Line 22/63 fix punctuation "F8-related arguments"

NVIDIA H100 GPU introduced support for a new datatype, FP8 (8-bit floating point), enabling higher throughput of matrix multiplies and convolutions. NeMo Framework uses the NVIDIA [TransformerEngine](https://github.com/NVIDIA/TransformerEngine) (TE) to leverage speedups from FP8. The following table summarizes the FP8-related arguments that can be configured in NeMo ([example config setting](https://github.com/NVIDIA/NeMo/blob/2e1814c9f031ad2aeeebad44597365e97253d2c4/examples/nlp/language_modeling/conf/megatron_gpt_config.yaml/#L192-L200)). For a more detailed overview, refer to the TE [documentation](https://docs.nvidia.com/deeplearning/transformer-engine/user-guide/index.html), specifically the FP8 [format](https://docs.nvidia.com/deeplearning/transformer-engine/user-guide/api/common.html#transformer_engine.common.recipe.Format) and [recipe](https://docs.nvidia.com/deeplearning/transformer-engine/user-guide/api/common.html#transformer_engine.common.recipe.DelayedScaling).

File: docs/source/features/mixed_precision.rst [hunk 5]
```diff
@@ -27,25 +68,46 @@ NVIDIA H100 GPU introduced support for a new datatype, FP8 (8-bit floating point
...
      - Description
-   * - transformer_engine
-     - TE and related functionality can be enabled by setting this boolean argument to True. If this argument is not set to True, all subsequent arguments will be ignored.
    * - fp8
-     - Enables FP8 training. For transformer networks, the QKV, projection, FC1, and FC2 matrix multiplications are executed using the fourth-generation NVIDIA H100 Tensor Cores with FP8 support.
-   * - fp8_e4m3
-     - Training recipe format for FP8. Activations, weights, and gradient tensors use the E4M3 format.
-   * - fp8_hybrid
-     - Training recipe format for FP8. Activations and weight tensors use the E4M3 format, whereas gradient use the E5M2 format to satisfy the additional dynamic range requirement for backward tensors. This is the default setting.
+     - Training recipe format for FP8: 'hybrid' or 'e4m3'. 'hybrid' is set by default. With 'hybrid', activations and weight tensors use the E4M3 format, whereas gradient use the E5M2 format to satisfy the additional dynamic range requirement for backward tensors.
```
Comment (jgerh) [inline docs/source/features/mixed_precision.rst:72]: suggested revision

The training recipe format for FP8 can be set to either 'hybrid' or 'e4m3', with 'hybrid' being the default. In the 'hybrid' format, activations and weight tensors use the E4M3 format, while gradients use the E5M2 format to meet the additional dynamic range requirements for backward tensors.

File: docs/source/features/mixed_precision.rst [hunk 6]
```diff
@@ -27,25 +68,46 @@ NVIDIA H100 GPU introduced support for a new datatype, FP8 (8-bit floating point
...
    * - fp8_margin
      - The scaling factor for FP8 tensors can be shifted by a factor of $2 ^ {margin}$ using this argument.
    * - fp8_amax_history_len
      - Window size for amax history. The window size determines how many instances of the most recent absolute max values (amaxes) are stored per tensor.
```
Comment (jgerh) [inline docs/source/features/mixed_precision.rst:76]: revise

 - The window size for amax history. The window size determines how many instances of the most recent absolute max values (amaxes) are stored per tensor.

File: docs/source/features/mixed_precision.rst [hunk 7]
```diff
@@ -27,25 +68,46 @@ NVIDIA H100 GPU introduced support for a new datatype, FP8 (8-bit floating point
...
    * - fp8_amax_compute_algo
      - The choice between “max” and “most_recent” specifies how to select an amax value from the given history.
-   * - reduce_amax
-     - Indicates whether or not to perform an allreduce on the amax (absolute max) values for the FP8 tensors. Since the amax is directly used to compute the scaling factor for FP8 tensors, setting this argument ensures that the scaling factors for a tensor remain synchronized across devices in multi-GPU training configurations.
    * - fp8_params
      - Indicates whether to store module-level parameters in FP8. Enabling this option can reduce memory consumption by eliminating the need to store a copy of weights in higher precision for cases where these weights are externally maintained, such as master parameters in the optimizer. For more information, refer to the `fp8_model_init <https://docs.nvidia.com/deeplearning/transformer-engine/user-guide/api/pytorch.html#transformer_engine.pytorch.fp8_model_init>`_ API in TE.
 
+Usage
```
Comment (jgerh) [inline docs/source/features/mixed_precision.rst:82]: Replace the "Usage" heading with "Implement FP8 Training" or "Apply FP8 Training".

Implement FP8 Training
^^^^^^^^^^^^^^^

File: docs/source/features/optimizations/attention_optimizations.rst [hunk 8]
```diff
@@ -56,22 +56,35 @@ To use MQA or GQA in the NeMo Framework, adjust the ``num_query_groups`` paramet
 1. **For Multi-query Attention (MQA)**:
    - Set ``num_query_groups`` to `1` to treat all attention heads as a single group.
```
Comment (jgerh) [inline docs/source/features/optimizations/attention_optimizations.rst:57]: Line 14/14 add "and"

The tiling technique decomposes the inputs based on the shared memory size and calculates the softmax one tile at a time. Instead of working on the entire query, key, and value tensors at once, it makes several passes at these tensors and then combines the results in a subsequent step.

Line 18/18 delete "and"

Flash attention lowers the memory footprint and computational complexity from quadratic to linear, greatly extending the range of sequence length allowed in large language models.

Line 25/25 fix capitalization "Framework"

In the NeMo Framework, flash attention is supported through `Transformer Engine <https://github.com/NVIDIA/TransformerEngine/tree/main>`_, 

Line 57/57 remove dash character in "- Set ``num_query_groups`` to `1` to treat all attention heads as a single group." It's not needed.

File: docs/source/features/parallelisms.rst [hunk 9]
```diff
@@ -41,54 +41,83 @@ Enable Data Parallelism
...
-
-NeMo Framework uses `PyTorch's FSDP interface <https://pytorch.org/tutorials/intermediate/FSDP_tutorial.html>`_ to shard LLM model states, flattening the parameters of each transformer layer and partitioning them across data-parallel GPUs.
-FSDP introduces collective operations across data-parallel GPUs, including all-gather for parameter computation and reduce-scatter for parameter gradients.
-The all-gather operation occurs during both the network forward and back-propagation phases, while the gradient reduce-scatter operation happens only during back-propagation.
-These FSDP communications are overlapped with transformer layer computations.
-
-Setting ``fsdp=true`` enables FSDP.
-The mixed precision recipe can be set by ``precision`` knob, which determines both the computation and communication precisions.
-Also, one can use ``grad_reduce_dtype`` to override the gradient reduction precision specifically.
+Currently, NeMo Framework supports optimizer distribution only for Megatron-Core adam distributed optimizer.
```
Comment (jgerh) [inline docs/source/features/parallelisms.rst:44]: revise

Currently, NeMo Framework supports optimizer distribution only for the Megatron Core Adam distributed optimizer.

File: docs/source/nlp/nemo_megatron/batching.rst [hunk 10]
```diff
@@ -5,17 +5,37 @@ Batching
 
 Batch size is one of the first parameters you should play with. For efficiency and convergence reasons we recommend you first try maximizing your batch size per GPU so that your GPU RAM usage is maximized.
```
Comment (jgerh) [inline docs/source/nlp/nemo_megatron/batching.rst:6]: suggested revision

Batch size is one of the first parameters you should adjust. For efficiency and convergence, we recommend first maximizing your batch size per GPU to fully utilize your GPU RAM.

File: docs/source/nlp/nemo_megatron/batching.rst [hunk 11]
```diff
@@ -5,17 +5,37 @@ Batching
...
 
-NeMo Megatron uses the following concepts.
+NeMo Framework uses the following concepts:
```
Comment (jgerh) [inline docs/source/nlp/nemo_megatron/batching.rst:8]: revise

NeMo Framework uses the following parameters:

File: docs/source/nlp/nemo_megatron/batching.rst [hunk 12]
```diff
@@ -5,17 +5,37 @@ Batching
...
 
-*Micro batch size* is the number of examples per data parallel rank. It is controlled by ``model.micro_batch_size`` parameter.
+===========================  ==================================================================================================================================
+Parameter                    Description
+===========================  ==================================================================================================================================
+Micro Batch Size             The number of examples per data parallel rank.
+Global Batch Size            ``global batch size = micro_batch_size * data_parallel_size * gradient_accumulation_steps``. For details on ``data_parallel_size`` see `this page <https://github.com/NVIDIA/NeMo/blob/main/docs/source/features/parallelisms.rst>`_.
```
Comment (jgerh) [inline docs/source/nlp/nemo_megatron/batching.rst:14]: revise

The global batch size is calculated as: global batch size = micro_batch_size * data_parallel_size * gradient_accumulation_steps.  For details on ``data_parallel_size``,  see `this page <https://github.com/NVIDIA/NeMo/blob/main/docs/source/features/parallelisms.rst>`_.

File: docs/source/nlp/nemo_megatron/batching.rst [hunk 13]
```diff
@@ -5,17 +5,37 @@ Batching
...
+Gradient Accumulation        Supports training with large batch sizes with fixed memory footprint at the cost of additional compute. ``accumulate_grad_batches`` is automatically handled by PyTorch Lightning.
```
Comment (jgerh) [inline docs/source/nlp/nemo_megatron/batching.rst:15]: suggested revision

This parameter supports training with large batch sizes while maintaining a fixed memory footprint, though it requires additional compute. The ``accumulate_grad_batches`` parameter is automatically managed by PyTorch Lightning.

File: docs/source/nlp/nemo_megatron/batching.rst [hunk 14]
```diff
@@ -5,17 +5,37 @@ Batching
...
+===========================  ==================================================================================================================================
 
-*Global batch size* = micro_batch_size * data_parallel_size * gradient_accumulation_steps. For details on ``data_parallel_size`` see :ref:`parallelisms` section, but typically it is equal to the number of GPUs being used.
-Global batch size is controlled by ``model.global_batch_size`` parameter. 
+Usage
```
Comment (jgerh) [inline docs/source/nlp/nemo_megatron/batching.rst:18]: suggest a heading revision to be more descriptive and add some introductory text to the code snippet

Set the Batching Parameters

The following example shows how to set up a pretraining recipe and batching parameters for a LLaMA-3 8B model.

File: docs/source/nlp/nemo_megatron/gpt/gpt_training.rst [hunk 15]
```diff
@@ -106,130 +102,24 @@ The memory map format makes training more efficient, especially with many nodes
...
-	exp_manager.checkpoint_callback_params.mode=min \
-	exp_manager.checkpoint_callback_params.always_save_nemo=False
-
-
-Next, you can launch Tensorboard to monitor training, as follows:
-
-.. code-block:: bash
-
-    tensorboard --logdir nemo_experiments --bind_all
+Once you have prepared training data, tokenizer, and recipe you are ready to train the model. You can follow `this tutorial <https://github.com/NVIDIA/NeMo/blob/main/examples/llm/pretrain/README.md#run-pre-training-with-a-default-recipe>`_
```
Comment (jgerh) [inline docs/source/nlp/nemo_megatron/gpt/gpt_training.rst:115]: revise

Once you have prepared the training data, tokenizer, and recipe, you are ready to train the model.

File: docs/source/nlp/nemo_megatron/gpt/gpt_training.rst [hunk 16]
```diff
@@ -106,130 +102,24 @@ The memory map format makes training more efficient, especially with many nodes
...
-	exp_manager.checkpoint_callback_params.always_save_nemo=False
-
-
-Next, you can launch Tensorboard to monitor training, as follows:
-
-.. code-block:: bash
-
-    tensorboard --logdir nemo_experiments --bind_all
+Once you have prepared training data, tokenizer, and recipe you are ready to train the model. You can follow `this tutorial <https://github.com/NVIDIA/NeMo/blob/main/examples/llm/pretrain/README.md#run-pre-training-with-a-default-recipe>`_ 
+to train a model using an existing recipe or `this one <https://github.com/NVIDIA/NeMo/blob/main/examples/llm/pretrain/README.md#create-and-run-a-custom-recipe>`_ to train a model with a custom recipe.
```
Comment (jgerh) [inline docs/source/nlp/nemo_megatron/gpt/gpt_training.rst:116]: revise

To train a model using an existing recipe, refer to this `tutorial <https://github.com/NVIDIA/NeMo/blob/main/nemo/collections/llm/recipes>`_ . To train a model with a custom recipe, follow this `guide <https://github.com/NVIDIA/NeMo/blob/main/examples/llm/pretrain/README.md#create-and-run-a-custom-recipe>`_.

File: docs/source/nlp/nemo_megatron/gpt/gpt_training.rst [hunk 17]
```diff
@@ -106,130 +102,24 @@ The memory map format makes training more efficient, especially with many nodes
     --workers=32
 
 
-Train a GPT-Style Model
-~~~~~~~~~~~~~~~~~~~~~~~
+Create Custom Train Recipe
```
Comment (jgerh) [inline docs/source/nlp/nemo_megatron/gpt/gpt_training.rst:105]: revise

Create a Custom Training Recipe

File: docs/source/nlp/nemo_megatron/gpt/gpt_training.rst [hunk 18]
```diff
@@ -4,10 +4,6 @@ GPT Model Training
 The Generative Pre-trained Transformer (GPT) is a decoder-only Transformer model. This section demonstrates how to train a GPT-style model with NeMo.
 
 
-
-
-
-
 .. note::
     This example is best completed using the latest NeMo Framework Training container `<https://catalog.ngc.nvidia.com/orgs/nvidia/containers/nemo>`_.
```
Comment (jgerh) [inline docs/source/nlp/nemo_megatron/gpt/gpt_training.rst:9]: Line 14/10 revise heading

Download and Preprocess the Data

**Review by jgerh — APPROVED** [review]
Reviewed changes and approved.
